### PR TITLE
UCS/BITMAP/TEST: Implement dynamic bitmap data structure

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Implementation of a bitmap that can grow as needed

### CODE REVIEW

File: src/ucs/datastruct/bitmap.h [hunk 1]
```diff
@@ -661,6 +669,181 @@ ucs_bitmap_bits_is_zero(const ucs_bitmap_word_t *bits, size_t num_words)
...
+    UCS_BITMAP_CHECK_INDEX(bit_index, num_words, <=);
+    return ucs_bitmap_bits_popcount(bits, last_word) +
+           ((mask == 0) ? 0 : ucs_popcount(bits[last_word] & mask));
+}
+
+
+/* Helper function to copy between bitmap bit arrays. Destination size must
+   be at lease the source size */
+static UCS_F_ALWAYS_INLINE void
+ucs_bitmap_bits_copy(const ucs_bitmap_word_t *src_bits, size_t src_num_words,
```
Comment (iyastreb) [inline src/ucs/datastruct/bitmap.h:705]: I noticed that in other functions destination params come first, followed by source ones.
In this one the order is the opposite.
Maybe it makes sense to maintain the same order `(dst, src)` in all similar functions to avoid confusion?

File: test/gtest/ucs/test_bitmap.cc [hunk 2]
```diff
@@ -352,6 +352,22 @@ class test_dynamic_bitmap : public ucs::test {
...
+
+    static void set(ucs_dynamic_bitmap_t &bitmap,
+                    const std::initializer_list<size_t> &offsets)
+    {
+        for (auto offset : offsets) {
+            ucs_dynamic_bitmap_set(&bitmap, offset);
+        }
+    }
+
+    void expect_bits(const std::initializer_list<size_t> &offsets)
```
Comment (iyastreb) [inline test/gtest/ucs/test_bitmap.cc:364]: IMO it would be better to have static function that takes bitmap by const reference, rather than using member `m_bitmap`. This way it's more reusable.
If you prefer to keep it as is, then this function may be const:
`void expect_bits(const std::initializer_list<size_t> &offsets) const`

File: src/ucs/datastruct/bitmap.h (hunk 1 above)
Comment (yosefe) [reply src/ucs/datastruct/bitmap.h:705]: right. but actually this function can be moved to a later PR since it's needed only by static_bitmap

File: test/gtest/ucs/test_bitmap.cc (hunk 2 above)
Comment (yosefe) [reply test/gtest/ucs/test_bitmap.cc:364]: changed to const

File: src/ucs/datastruct/bitmap.h [hunk 3]
```diff
@@ -661,6 +669,181 @@ ucs_bitmap_bits_is_zero(const ucs_bitmap_word_t *bits, size_t num_words)
...
+    ucs_bitmap_word_t mask = ucs_bitmap_word_bit_mask(bit_index) - 1;
+
+    UCS_BITMAP_CHECK_INDEX(bit_index, num_words, <=);
+    return ucs_bitmap_bits_popcount(bits, last_word) +
+           ((mask == 0) ? 0 : ucs_popcount(bits[last_word] & mask));
+}
+
+
+/* Helper function to copy between bitmap bit arrays. Destination size must
+   be at lease the source size */
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucs/datastruct/bitmap.h]: will fix in next PR (removed the whole function)

**Review by iyastreb — APPROVED** [review]

**Comment by yosefe** [comment]
@gleon99 can you pls review as well?

File: test/gtest/ucs/test_bitmap.cc [hunk 4]
```diff
@@ -368,3 +384,105 @@ UCS_TEST_F(test_dynamic_bitmap, test_get_set_reset)
 
     EXPECT_TRUE(ucs_dynamic_bitmap_is_zero(&m_bitmap));
 }
+
```
Comment (tvegas1) [inline test/gtest/ucs/test_bitmap.cc:387]: maybe we need to add tests for at least some of the `ucs_bitmap_bits_*()` functions?

Comment (yosefe) [reply test/gtest/ucs/test_bitmap.cc:387]: they should all be tested by static_bitmap/dynamic_bitmap, since they are internal helper function IMO they don't deserve a separate testing..

**Review by tvegas1 — APPROVED** [review]

File: test/gtest/ucs/test_bitmap.cc [hunk 5]
```diff
@@ -368,3 +384,105 @@ UCS_TEST_F(test_dynamic_bitmap, test_get_set_reset)
...
+    ucs_dynamic_bitmap_t bitmap2;
+    ucs_dynamic_bitmap_init(&bitmap2);
+
+    set(m_bitmap, {1, 3, 4});
+    set(bitmap2, {2, 3, 4, 5});
+
+    ucs_dynamic_bitmap_xor_inplace(&m_bitmap, &bitmap2);
+    expect_bits({1, 2, 5});
+
+    ucs_dynamic_bitmap_cleanup(&bitmap2);
```
Comment (gleon99) [inline test/gtest/ucs/test_bitmap.cc:487]: Maybe add (at least some basic sanity) scenarios for testing "zero-padding" in the various operations?

File: src/ucs/datastruct/bitmap.h [hunk 6]
```diff
@@ -661,6 +669,164 @@ ucs_bitmap_bits_is_zero(const ucs_bitmap_word_t *bits, size_t num_words)
...
+           ((mask == 0) ? 0 : ucs_popcount(bits[last_word] & mask));
+}
+
+
+/* Helper function to set the bitmap array to a mask up to the given index */
+static UCS_F_ALWAYS_INLINE void ucs_bitmap_bits_mask(ucs_bitmap_word_t *bits,
+                                                     size_t num_words,
+                                                     size_t bit_index)
+{
+    size_t last_word = bit_index / UCS_BITMAP_BITS_IN_WORD;
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucs/datastruct/bitmap.h:707]: cannot use const here since it's incremented in the function

File: src/ucs/datastruct/bitmap.h [hunk 7]
```diff
@@ -661,6 +669,164 @@ ucs_bitmap_bits_is_zero(const ucs_bitmap_word_t *bits, size_t num_words)
...
+    size_t word_index;
+
+    UCS_BITMAP_CHECK_DST_NUM_WORDS(dst_num_words, src_num_words);
+
+    for (word_index = 0; word_index < src_num_words; ++word_index) {
+        dst_bits[word_index] = ~src_bits[word_index];
+    }
+
+    if (dst_num_words > src_num_words) {
+        /* Set remaining bits in destination to 1 */
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucs/datastruct/bitmap.h:739]: not sure what you mean here?

**Review by gleon99 — APPROVED** [review]
